src/transaction/log_recovery.c

```diff
@@ -1049,6 +1049,74 @@ log_recovery_finish_transactions (THREAD_ENTRY * const thread_p)
   // final recovery phase flag is set in the caller
 }
 
+/* log_recovery_analysis_from_transaction_table_snapshot - perform recovery for a passive transaction server
+ *                  starting from a [recent] transaction table snapshot relayed via log from the active
+ *                  transaction server
+ *
+ * most_recent_transaction_table_snapshot_lsa (in): the lsa where a record containing, as payload, the packed contents
+ *                                of a recent transaction table snapshot; starting from that snapshot, analyze the log
+ *                                to construct an actual starting transaction table
+ */
+void
+log_recovery_analysis_from_transaction_table_snapshot (THREAD_ENTRY * thread_p,
+						       log_lsa most_recent_transaction_table_snapshot_lsa)
+{
+  assert (!most_recent_transaction_table_snapshot_lsa.is_null ());
+  assert (LOG_CS_OWN_WRITE_MODE (thread_p));
+
+  // analysis changes the transaction index and leaves it in an indefinite state
+  // therefore reset to system transaction index afterwards;
+  // first make sure we're executing on the system thread
+  const int sys_tran_index = LOG_FIND_THREAD_TRAN_INDEX (thread_p);
+  assert (sys_tran_index == LOG_SYSTEM_TRAN_INDEX);
+
+  log_reader lr (LOG_CS_SAFE_READER);
+  int log_page_read_err = lr.set_lsa_and_fetch_page (most_recent_transaction_table_snapshot_lsa);
+  if (log_page_read_err != NO_ERROR)
+    {
+      logpb_fatal_error (thread_p, true, ARG_FILE_LINE,
+			 "log_initialize_passive_tran_server: error reading transaction table snapshot log page");
+      return;
+    }
+
+  // *INDENT-OFF*
+  // copy because, maybe, the next add will advance to a new page
+  const log_rec_header log_rec_hdr = lr.reinterpret_copy_and_add_align<log_rec_header> ();
+  // *INDENT-ON*
+  assert (log_rec_hdr.type == LOG_TRANTABLE_SNAPSHOT);
+
+  lr.advance_when_does_not_fit (sizeof (LOG_REC_TRANTABLE_SNAPSHOT));
+  // *INDENT-OFF*
+  const LOG_REC_TRANTABLE_SNAPSHOT *const log_rec = lr.reinterpret_cptr<LOG_REC_TRANTABLE_SNAPSHOT>();
```

Comment:
Copy because the page may advance.
```suggestion
  const LOG_REC_TRANTABLE_SNAPSHOT log_rec = lr.reinterpret_copy_and_add_align<LOG_REC_TRANTABLE_SNAPSHOT>();
```

Reply:
Done